refactor: Replace Addict with Guardian

pmrukot opened the pull request:
**Summary**
Replaces Addict with Guardian. Basically changes whole authentication model.

**Related issues**
https://github.com/forcecraft/aion/issues/39

**Test plan**
Manual + additional unit tests. Register and login are not integrated in Elm as a part of single page app.

**TODO**
- [x] Delete Addict (libraries + templates)
- [x] Setup Guardian
- [x] Add login page
- [x] Add registration page
- [x] Add logout
- [x] Persist token in local storage
- [x] Add token while making request for resources
- [x] Update channel token


### NOTE(mrapacz|12-10-2017):
What still needs to be done before merging:
 - [x] A proper review
 - [x] Backend unit tests for registration and all the other stuff that appeared
 - [x] Moving the keys to envs and replacing the hardcoded values with `System.get_env`

Commit:
fix: Remove addict

Commit:
Initial version

mrapacz commented on aion/config/dev.exs:
```diff
@@ -26,18 +19,21 @@ config :aion, Aion.Endpoint,
+
+config :guardian, Guardian,
+  allowed_algos: ["HS512"],
+  secret_key: "@11H41LW35tT3x45",
```
:laughing:

pmrukot replied:
Thats super experimental PR, please do not laugh 😠

pmrukot commented:
Added TODO section for easier tracking of a PR

mrapacz commented:
Note that this PR's main goal is to swap the libraries.

+ Adding logout is not an essential part of this PR, it may actually be easier to push it as a separate change (we were able to do without it so far, which means we should be able to do without it in this PR too). 
 + Adding login/registration page is not essential (at least not both of them at the same time). So far we  were able to use only registration page to log in and test that everything works correctly and that may be enough to verify that PR changes are alright. In a separate PR we could extend the registration page with the login feature. If you introduce two different pages right now then it will be much more complex to merge them into one that would offer both registration and login later.

pmrukot commented:
I'll make login/register page as one view. Idea of this PR is to make it work just as it was before. I agree that logout is not necessary but our users need to be able to register/login just as before.

mrapacz commented:
This PR should contain whatever changes are needed to prove that addict was removed totally and guardian was successfully introduced. From that point we can begin to work paralelly on issues connected to that - logout for example (users didn't have an option to do that so far).

Commit:
Make sockets auth work

Commit:
Resolve get_user function ambiguity

Commit:
Prepare for Registration form

Commit:
Initial register form

Commit:
Working version of registration

Commit:
Fetch resources after login

Commit:
Make local storage work

Commit:
Try to init socket connection when app starts

Commit:
Make /api/me work again, prototype of request with tokens

Commit:
Make authenticated request to api/me

Commit:
Delete unused code

Commit:
Remove temp function

Commit:
Add logout

Commit:
Prevent errors on /api/me

Commit:
Protect sources from unauthorised request, add token to requests on Elm side

Commit:
Merge branch 'master' into refactor/replace-addict-with-guardian

Commit:
Make credo happy

pmrukot commented:
Everything is done, I think this is enough for first version. Im not sure about the case when the token expires but lets leave this case for now. Branch is also synchronised with master.

But I have another issues that are more important:
- need to verify if config is fine / add configuration for production
- need help with dialyzer
- `mix.test` is screaming something like:
```ex
== Compilation error on file lib/aion/endpoint.ex ==
** (UndefinedFunctionError) function Aion.Router.init/1 is undefined (module Aion.Router is not available)
    Aion.Router.init([])
    (plug) lib/plug/builder.ex:193: Plug.Builder.init_module_plug/3
    (plug) lib/plug/builder.ex:181: anonymous fn/4 in Plug.Builder.compile/3
    (elixir) lib/enum.ex:1755: Enum."-reduce/3-lists^foldl/2-0-"/3
    (plug) lib/plug/builder.ex:181: Plug.Builder.compile/3
    (phoenix) expanding macro: Phoenix.Endpoint.__before_compile__/1
    lib/aion/endpoint.ex:1: Aion.Endpoint (module)
```

So 2 important checks wont pass on CI (tests wont run and dialyzer also screams)

pmrukot commented:
New login change:
<img width="1176" alt="screen shot 2017-10-11 at 22 51 18" src="https://user-images.githubusercontent.com/13131966/31466213-bd4cd6c4-aed6-11e7-83ca-a8ba8642be9b.png">

New register page:
<img width="1190" alt="screen shot 2017-10-11 at 22 51 25" src="https://user-images.githubusercontent.com/13131966/31466214-bdfbb194-aed6-11e7-9ba5-a7184d3cf7a6.png">


The rest is working as before. Can't wait to actually split this into two apps/repositories.

pmrukot commented:
Start with manual tests, let's fix the CI and then start reviewing the code. Please checkout to this branch and if everything works for you as before.

mrapacz commented:
I'll read into that on Friday, I think if we merge it before next Wednesday, that's good. 
As for the configs, we cannot hardcode any values. They'll need to be imported from envs, so we'll probably need to add a key to travis' env and everyone will have to add a variable on their own dev, same for prod.

Commit:
Make tests run again

Commit:
Make tests work again

pmrukot commented:
Testes are fixed, dialzyer and deploy related stuff left (+ some manual tests of course)

Commit:
Add guardian setup to travis.exs

Commit:
Fix indentation

mrapacz commented:
I'm looking into Dialyzer now. The room_channel errors are to be added to `ignore_warnings`, because they won't work well until we add correct type specs for the whole module. This is to be done as #120 

I'm investigating the others, will edit the comment as soon as I get to know something.

mrapacz committed:
Update mix.lock

mrapacz committed:
Add room_channel warnings to dialyzer warning file

mrapacz committed:
Fix dialyzer reports

mrapacz commented:
I tried experimenting with different approaches in `mix.exs` configuration of dialyxir but to no avail.
[Here's the option I'm talking about](https://github.com/jeremyjh/dialyxir#dependencies) 

These are the warnings I'm experiencing currently:
```
:0: Unknown function 'Elixir.Comeonin.Bcrypt':checkpw/2
:0: Unknown function 'Elixir.Comeonin.Bcrypt':dummy_checkpw/0
:0: Unknown function 'Elixir.Comeonin.Bcrypt':hashpwsalt/1
:0: Unknown function 'Elixir.Guardian.Phoenix.Socket':sign_in/2
:0: Unknown function 'Elixir.Guardian.Plug':api_sign_in/2
:0: Unknown function 'Elixir.Guardian.Plug':claims/1
:0: Unknown function 'Elixir.Guardian.Plug':current_resource/1
:0: Unknown function 'Elixir.Guardian.Plug':current_token/1
:0: Unknown function 'Elixir.Guardian.Plug.EnsureAuthenticated':call/2
:0: Unknown function 'Elixir.Guardian.Plug.LoadResource':call/2
:0: Unknown function 'Elixir.Guardian.Plug.VerifyHeader':call/2
web/auth/serializer.ex:1: Callback info about the 'Elixir.Guardian.Serializer' behaviour is not available
```

They all concern the new deps connected to Guardian. Depending on the options I've tested, there were different results. Possibly the problem here is that some of these functions are not seen because they are deps of Guardian, not strictly our deps and it doesn't investigate that far with currect config. I think they can be safely added to ignore_warnings for now and we will have to fix that along with #101 


This one was fixed:

```
web/controllers/registration_controller.ex:6: Call to missing or unexported function 'Elixir.Aion.User':'get!'/2
``` 

but there are no automated tests to check if it's still working. The tests for new views *must* be implemented, they are the easiest way to verify that the endpoints will work as they should. As for now, my change does not break anything during the tests, which should not be case I suspect. 

Also, [Ecto.Schema docs](https://hexdocs.pm/ecto/Ecto.Schema.html) always use Repo like `Repo.get!(Model, id)`, never like `Model.get!(Model, id)`, so I'm assuming dialyzer was right to report the issue.

mrapacz committed:
Uncomment bundle.js

Commit:
Update dialyzer warnings

mrapacz commented on aion/config/travis.exs:
```diff
@@ -12,4 +12,11 @@ config :aion, Aion.Repo,
\ No newline at end of file
+  pool: Ecto.Adapters.SQL.Sandbox
+
+config :guardian, Guardian,
```
There should be one Guardian config only in config.exs, it should import a proper key from the environment (which will vary depending on the environment).

mrapacz commented on aion/mix.exs:
```diff
@@ -11,51 +11,39 @@ defmodule Aion.Mixfile do
-  ]
+     {:postgrex, ">= 0.0.0"},
+     {:simetric, "~> 0.1.0"},
+     {:exrm, "~> 1.0.8"},]
```
Move the bracket to a new line.

mrapacz commented on aion/test/controllers/answer_controller_test.exs:
```diff
@@ -6,9 +6,13 @@ defmodule Aion.AnswerControllerTest do
-    user = %{ email: "test@example.com", name: "something" }
-    conn = Plug.Test.init_test_session(build_conn(), %{current_user: user})
-    {:ok, conn: put_req_header(conn, "accept", "application/json")}
+    user = %Aion.User{ email: "test@example.com", name: "something", password: "2131231", id: 1 }
```
There should be no spaces after `{` and before `}`

mrapacz commented on aion/test/controllers/category_controller_test.exs:
```diff
@@ -6,9 +6,13 @@ defmodule Aion.CategoryControllerTest do
-    user = %{ email: "test@example.com", name: "something"}
-    conn = Plug.Test. init_test_session(build_conn(), %{current_user: user})
-    {:ok, conn: put_req_header(conn, "accept", "application/json")}
+    user = %Aion.User{ email: "test@example.com", name: "something", password: "2131231", id: 1 }
```
`{ x }`

mrapacz commented on aion/test/controllers/room_controller_test.exs:
```diff
@@ -6,9 +6,13 @@ defmodule Aion.RoomControllerTest do
-    user = %{ email: "test@example.com", name: "something" }
-    conn = Plug.Test. init_test_session(build_conn(), %{current_user: user})
-    {:ok, conn: put_req_header(conn, "accept", "application/json")}
+    user = %Aion.User{ email: "test@example.com", name: "something", password: "2131231", id: 1 }
```
`{ samehere }`

Commit:
Centralize config

mrapacz commented on aion/web/auth/auth.ex:
```diff
@@ -0,0 +1,43 @@
+    Keyword.fetch!(opts, :repo)
+  end
+
+  def call(conn, repo) do
```
This could be transformed to a nice pipeline I guess. I'm not sure if it's used at all actually (I cannot find any direct call of this function in our code). Maybe it's safe to delete, this has to be checked after writing tests for auth in general.

mrapacz commented on aion/web/auth/auth.ex:
```diff
@@ -0,0 +1,43 @@
+  def login_by_username_and_pass(conn, email, given_pass, opts) do
+    repo = Keyword.fetch!(opts, :repo)
+    user = repo.get_by(Aion.User, email: email)
+    cond do
```
I'm not a particular fan of implicit casting of data structures to booleans. It makes the code less legible (especially when there are no typespecs and you have no idea of what `user` is.

mrapacz replied:
I mean, maybe `case` would work better here?

pmrukot replied:
I  think it makes a lot of sense, is shorter and more practical to do so. I think its a common pattern in dynamic languages and I use it extensively in JS so I do not consider it something bad. And `repo.get_by` can either return `User` or `nil`.

mrapacz replied:
How does checkpw work? Will it return true|false everytime or can it raise an error or something? 

If it's the first case, then maybe [this kind of case statement](https://stackoverflow.com/questions/21075026/what-is-the-difference-between-cond-and-case) would be good here?

pmrukot replied:
It does return true/false and I think this is a correct way of handling such cases (the way its done now). This `case` statement does not seem to give any advantage comparing to current solution.

mrapacz replied:
Okay, let's leave it this way then

Commit:
Remove unused functions

Commit:
Reset location after login

mrapacz commented on aion/web/channels/room_channel.ex:
```diff
@@ -130,7 +130,8 @@ defmodule Aion.RoomChannel do
 
-  defp get_user(socket) do
-    socket.assigns.current_user.name
+  defp get_username(socket) do
```
In general I'm not a fan of lack of consequent in naming things. We have `user.name` and the function is named `get_username`. Maybe we should rename the user's field to `username`?

Also, this may be fixed the moment we remove `email` from the form and introduce `Login` (unless we want to introduce some form of email confirmation).

pmrukot replied:
is `get_user_name` a good middle ground?

mrapacz replied:
Yes, let's stay with that for now.

mrapacz commented on aion/web/channels/user_socket.ex:
```diff
@@ -10,15 +11,18 @@ defmodule Aion.UserSocket do
-    case Token.verify(socket, "user", token, max_age: 86_400) do
-      {:ok, user_id} ->
-        socket = assign(socket, :current_user, Repo.get!(User, user_id))
+  def connect(%{"token" => jwt}, socket) do
```
I know that Guardian is using JWT as the tokens, but maybe the name `token` wasn't that bad? You don't really have to care that much what kind of token that is (or this could be explained in typespecs).

mrapacz commented on aion/web/channels/user_socket.ex:
```diff
@@ -10,15 +11,18 @@ defmodule Aion.UserSocket do
+    case sign_in(socket, jwt) do
+      {:ok, auth_socket, guardian_params} ->
+        {:ok, auth_socket}
+      _ ->
```
So we're always returning {:ok, some_kind_of_socket}? Is there a case when we would return `:error`? (The original typespecs for `Phoenix.Socket` allow such a thing, so I'm wondering what is the use case for that.

pmrukot replied:
Probably need to think about better error handling but Im not sure if that would pay off. The only case is where we there is either no token provided or it expired. For now I would leave it since frontend also does not have a solution for refreshing expired token.

I would make an issue of that - to improve the way we handle errors related to authentication. For now I will just return `:error` as suggested in docs: https://hexdocs.pm/phoenix/Phoenix.Socket.html#callbacks

mrapacz replied:
Yes, that's exactly what I was talking about - the callback saying there may be a case for throwing an error. You should probably make it return error in those cases then and also create the said issue.

mrapacz commented on aion/web/controllers/answer_controller.ex:
```diff
@@ -52,4 +54,10 @@ defmodule Aion.AnswerController do
+  def unauthenticated(conn, _params) do
+    conn
+    |> put_status(401)
+    |> render("error.json", message: "Authentication required")
```
I'm thinking about putting common functions responsible for displaying errors to one module, possibly located in `web/controllers/errors.ex` where you would have a function `render_unauthenticated_error`. 

This way you would not have to hardcode the same string message and error.json multiple times and the pipeline would be somewhat more legible (instead of `render(..)` you would have `|> render_unauthenticated_error()`.

jtkpiotr replied:
I agree but I would prefer to put it in authorization related module

mrapacz replied:
This is going to be a one function only. I think it could be better if we just stored all status-related things used across all controllers in one file around the controllers. There is a reason why the unauthenticated function lies in the controller implementation and I guess anyone from outside the project trying to see where they get a certain status would look in the controller area. 

I would agree to put it in auth area if it was auth only thing, but there are also other error statuses and splitting them to be in different areas would make it harder to find certain implementations. We're going to have to create the `errors.ex` or something like that anyway for cases like `unprocessable entity` etc. Unless there is a module in authentication area where this function would fit in perfectly, I would opt for putting all common error rendering functions in one module.

mrapacz commented on aion/web/controllers/category_controller.ex:
```diff
@@ -52,4 +54,10 @@ defmodule Aion.CategoryController do
+  def unauthenticated(conn, _params) do
+    conn
+    |> put_status(401)
+    |> render("error.json", message: "Authentication required")
```
^ `errors.ex`

mrapacz commented on aion/web/controllers/question_controller.ex:
```diff
@@ -52,4 +54,10 @@ defmodule Aion.QuestionController do
+  def unauthenticated(conn, _params) do
+    conn
+    |> put_status(401)
+    |> render("error.json", message: "Authentication required")
```
^ `errors.ex`

mrapacz commented on aion/web/controllers/room_category_controller.ex:
```diff
@@ -52,4 +54,10 @@ defmodule Aion.RoomCategoryController do
+  def unauthenticated(conn, _params) do
+    conn
+    |> put_status(401)
+    |> render("error.json", message: "Authentication required")
```
^ `errors.ex`

mrapacz commented on aion/web/controllers/user_controller.ex:
```diff
@@ -1,9 +1,17 @@
+  def unauthenticated(conn, _params) do
+    conn
+    |> put_status(401)
+    |> render("error.json", message: "Authentication required")
```
^ `errors.ex`

Commit:
Don't hardcode url

mrapacz commented on aion/web/controllers/registration_controller.ex:
```diff
@@ -0,0 +1,24 @@
+        |> render("show.json", user: user)
+      {:error, changeset} ->
+        conn
+        |> put_status(:unprocessable_entity)
```
This could also be moved to `errors.ex` like:
```
conn
|> render_unprocessable_entity()
```
or you could even get rid of the pipeline here

mrapacz commented on aion/web/controllers/session_controller.ex:
```diff
@@ -0,0 +1,29 @@
+        new_conn = Plug.api_sign_in(conn, conn.assigns[:current_user])
+        jwt = Plug.current_token(new_conn)
+        {:ok, claims} = Plug.claims(new_conn)
+        exp = Map.get(claims, "exp")
```
You can go with `exp = claims.exp` here

pmrukot replied:
`claims["exp"]`*

mrapacz replied:
Yeah right, _'dis ain't no atom dude'_

Commit:
Add notification when login fails

mrapacz commented on aion/web/elm/src/Auth/Api.elm:
```diff
@@ -0,0 +1,75 @@
+
+
+authenticate : String -> Decode.Decoder a -> Value -> Http.Request a
+authenticate url dec credentials =
```
Maybe `decoder` would be better here? I had no idea what it is until I looked at type specs

mrapacz commented on aion/web/elm/src/Auth/Models.elm:
```diff
@@ -0,0 +1,61 @@
+-- login form section
+
+
+type alias LoginForm =
```
:+1:

mrapacz commented on aion/web/elm/src/Auth/Models.elm:
```diff
@@ -0,0 +1,61 @@
+    Forms.Form
+
+
+loginFormPossibleFields : List String
```
Isn't `loginFormFields` enough? I mean when you say that 'hey, these are login form fields', that kinda makes it clear that they are the possible options, too.

mrapacz commented on aion/web/elm/src/Auth/Notifications.elm:
```diff
@@ -0,0 +1,19 @@
+
+registrationSuccessfulToast : ( Model, Cmd Msg ) -> ( Model, Cmd Msg )
+registrationSuccessfulToast =
+    addToast (Toasty.Defaults.Success "Success!" "Account created successfuly - proceed to Login.")
```
`successfully` and `login` (or `Login Page`)

mrapacz replied:
^ otherwise it's not clear if login is a verb or a noun

Commit:
Remove wrong import

mrapacz commented on aion/web/elm/src/Auth/Views.elm:
```diff
@@ -0,0 +1,149 @@
+authView : Model -> Html Msg
+authView model =
+    div [ class "auth-container" ]
+        [ case model.authData.displayLoginInsteadOfRegistration of
```
How about:
 + `loginPageEnabled` 
 + `registrationPageHidden`
 + `displayLoginPage`? 

I think writing `displayAinsteadOfsomethingElse` is an overkill. A more functional approach would actually be to go like this:
```
type UnauthenticatedToggle = Login | Registration 
```
(cannot think of a better name for the type) and you would just do a 
```
case model.authData.unauthenticatedToggle (or unauthenticatedViewToggle or a better naming if you have an idea for that) of
    Login -> 
    Registration ->
``` 
^ look 100 times better imo. If you really want to check if something is true or false, maybe it's better to do it with an if statement. 

But I think defining a type and changing it to pattern matching is much more verbose and clearer - imagine a situation when we introduce the third page - register with a teacher's code that would instantly give you a teacher's permissions, it would just mean introducing third subtype like `Teacher` and extending the pattern matching, nothing else).

mrapacz commented on aion/web/elm/src/Auth/Views.elm:
```diff
@@ -0,0 +1,149 @@
+
+            False ->
+                registrationFormView model.authData.registrationForm
+        , br [] []
```
I guess normally you just deal with these things by setting up margins and stuff, right? I'm not gonna delay the merge with fixes like this, but I think there should be a follow-up issue for things like this maybe?

pmrukot replied:
Yeah probably, didn't want to add css to this PR. Its enough code.

mrapacz replied:
Ack, we should create a follow-up issue for that anyway so that we don't forget to fix this stuff later.

mrapacz commented on aion/web/elm/src/Auth/Views.elm:
```diff
@@ -0,0 +1,149 @@
+-- auth form switch section
+
+
+authFormSwitch : String -> Html Msg
```
Maybe `authFormToggle` would be better? Or `authFormToggleButton`.

mrapacz commented on aion/web/elm/src/Auth/Views.elm:
```diff
@@ -0,0 +1,149 @@
+
+
+
+-- login form section
```
Also all the other files in other modules are named `View.elm`, not `Views`

mrapacz commented on aion/web/elm/src/Auth/Views.elm:
```diff
@@ -0,0 +1,149 @@
+        ]
+
+
+passwordRegisterFormElement : Forms.Form -> Html Msg
```
All these formElement functions look the same, they just differ in the labels. The labels could be moved out and you could just define a function `formElement` which would deal with all of that code above and define emailRegisterFormElement as 
```
emailRegisterFormElement : Forms.Form -> RegistrationFormInput -> ErrorMessage -> Html Msg
emailRegisterFormElement form RegistrationFormInputs.Email RegistrationErrors.EmailErrorMessage = ...
```

I think it would take roughly 50 lines of code from this file?

mrapacz replied:
^ also email and email error would just be parts of a `type` which would allow you to get rid of the hardcoded strings :)

pmrukot replied:
I think making forms too abstract will make it harder to make adjustments later. It may turn out that we need to change markup of one field and we are forced to refactor whole function and we are not flexible at all. I think having html markup at low abstraction level is good thing, especially in early development of this parts of code.

mrapacz replied:
Okay, we may leave this for the #56

mrapacz commented on aion/web/elm/src/General/Constants.elm:
```diff
@@ -34,3 +32,37 @@ profileCardConfig location =
+
+loginFormMsg : String
+loginFormMsg =
+    "Don't have an account? Click here to register"
```
`.` at the end.

mrapacz commented on aion/web/elm/src/General/Constants.elm:
```diff
@@ -34,3 +32,37 @@ profileCardConfig location =
+
+registerFormMsg : String
+registerFormMsg =
+    "Already have an account? Click here to login"
```
Also `.` at the end.

mrapacz commented on aion/web/elm/src/General/Models.elm:
```diff
@@ -34,41 +36,48 @@ type alias Model =
             Navbar.initialState NavbarMsg
+
+        token =
+            case String.isEmpty flags.token of
```
How about 
```
case flags.token of
    "" -> Nothing
    token -> Just token
```

pmrukot replied:
`token` naming ambiguity. Will not get much better.

mrapacz commented on aion/web/elm/src/Panel/Api.elm:
```diff
@@ -1,25 +1,74 @@
+createQuestionWithAnswersRequest url token decoder body =
+    Http.request
+        { method = "POST"
+        , headers = [ Http.header "Authorization" ("Bearer " ++ token) ]
```
You could DRY out the code in the following couple of functions. For example you could add a function named `createRequest`:
```
createRequest url token decoder body = 
```
Notice that the following functions are _exactly_ the same in the implementation, the only different is the Decoder type.

pmrukot replied:
Would also need to add a `method` field. How can I make a function that takes concrete decoder type more generic?

mrapacz replied:
For the method itself you can always define 
```
type requestMethod = Get | Post
```
You will probably need to define something like
```
requestMethodToString : requestMethod -> String
requestMethodToString method = case method of
     Get -> "GET"
     Post -> "POST"
```
unless there's a better way to do it.

As for the generic types you can see how it's explained [here](https://stackoverflow.com/questions/38558964/understanding-generic-union-types-in-elm)

mrapacz replied:
@pmrukot if it doesn't get fixed now, it should be added as a follow-up issue (you can add it to changes listed in #56)

mrapacz commented on aion/web/elm/src/Update.elm:
```diff
@@ -21,11 +26,115 @@ import Task
+    Forms.updateFormInput form name value
+
+
+unwrapToken : Maybe String -> String
```
How about 
```
getToken maybeToken = 
    case maybeToken of
        Just token -> token
        Nothing -> ""
```

mrapacz commented on aion/web/elm/src/Update.elm:
```diff
@@ -21,11 +26,115 @@ import Task
+                oldDisplayLoginInsteadOfRegistration =
+                    oldAuthData.displayLoginInsteadOfRegistration
+            in
+                case oldDisplayLoginInsteadOfRegistration of
```
This will get hugely simplified when we introduce setters, so I'm not going to ask you to fix it now.

mrapacz commented on aion/web/elm/src/View.elm:
```diff
@@ -49,15 +51,38 @@ navbar location navbarState =
+        includeNavbar =
+            case currentRoute of
+                AuthRoute ->
+                    \x y z -> x
```
:1st_place_medal:

mrapacz replied:
Seriously, though, any idea for a better naming here? :smile:

mrapacz replied:
Plus if y and z are not needed then I suppose you could do `\betterName _ _ -> betterName`

mrapacz commented on aion/web/models/user.ex:
```diff
@@ -18,7 +20,26 @@ defmodule Aion.User do
-    |> validate_required([:name, :email, :encrypted_password])
+    |> cast(params, [:name, :email, :password])
+    |> validate_required([:name, :email, :password])
+    |> validate_length(:email, min: 3, max: 30)
```
We could move constants like this to a separate module or just out of this function and not hardcode them. (especially the latter one shouldn't be a lot of work, you could deal with that using the `module @attributes` which are conventional way of dealing with module constants in Elixir.

pmrukot replied:
This is out of scope of this PR.

mrapacz replied:
True

Commit:
Login user after register

mrapacz commented on aion/web/views/user_view.ex:
```diff
@@ -0,0 +1,17 @@
+  use Aion.Web, :view
+
+  def render("user.json", %{user: user}) do
+    if user != nil do
```
I'm still not sure about conventional way of dealing with nils, but [this answer](https://stackoverflow.com/a/37781248) and [Jose's answer](https://groups.google.com/d/msg/elixir-lang-talk/T6fg8h60Nnw/PodQeBGzGAAJ) suggest some solutions. I think the second one could actually be better (but it's just my personal opinion because I'm not a die hard fan of multiple function clauses). Here's how the first one could work: 
```
def render("user.json", %user: user}) when is_nil user, do: %{}
def render("user.json", %user: user}), do: Map.fetch(user, [:id, :name, :email])
```
Similarly to my suggestion ^ in the second one you should definitely use `Map.fetch/2`

pmrukot replied:
Would this work?
```ex
def render("user.json", %{user: nil}), do: %{}
def render("user.json", %{user: user}), do: Map.fetch(user, [:id, :name, :email])
```

pmrukot replied:
`Map.take/2`*

mrapacz replied:
yeah right, was looking at the docs and eventually used the wrong one
lgtm

mrapacz reviewed (changes requested):
Some of them are `to-do-now` issues, some of them are to be discussed. Fantastic job in general :clap: :fire: :fireworks: :sparkles: 

Too bad the PR is so big and the review takes so long :laughing:

Commit:
Fix post auth actions

mrapacz commented on aion/web/elm/src/Auth/Notifications.elm:
```diff
@@ -16,4 +16,13 @@ registrationErrorToast =
 registrationSuccessfulToast : ( Model, Cmd Msg ) -> ( Model, Cmd Msg )
 registrationSuccessfulToast =
-    addToast (Toasty.Defaults.Success "Success!" "Account created successfuly - proceed to Login.")
+    addToast (Toasty.Defaults.Success "Success!" "Account created successfuly.")
```
`successfully`

Commit:
Move socket initialisation to Socket.elm

jtkpiotr commented on aion/test/controllers/room_category_controller_test.exs:
```diff
@@ -5,9 +5,13 @@ defmodule Aion.RoomCategoryControllerTest do
-    user = %{ email: "test@example.com", name: "something" }
-    conn = Plug.Test. init_test_session(build_conn(), %{current_user: user})
-    {:ok, conn: put_req_header(conn, "accept", "application/json")}
+    user = %Aion.User{ email: "test@example.com", name: "something", password: "2131231", id: 1 }
```
this piece of code repeats many times ;) How about extracting it somewhere ;)

Commit:
Review adjustments

pmrukot commented:
I extracted unauthenticated function to `errors.ex` in views. Test setup is also extracted. As for making a generic request function I would need to put a little more effort into it. Apart from that most of the changes are applied, let me know what you think.

mrapacz commented on aion/test/test_helper.exs:
```diff
@@ -2,3 +2,18 @@ ExUnit.start
 
 Ecto.Adapters.SQL.Sandbox.mode(Aion.Repo, :manual)
 
+defmodule Aion.TestHelpers do
```
I guess the conventional way is to put it in the `support/` and name it with `_case`:
![image](https://user-images.githubusercontent.com/15965147/31625785-a24c1bb2-b2a7-11e7-879e-ddcc59b80d46.png)

mrapacz commented on aion/web/controllers/registration_controller.ex:
```diff
@@ -0,0 +1,22 @@
+        SessionController.create(conn, %{"email" => user.email, "password" => user.password})
+      {:error, changeset} ->
+        conn
+        |> put_status(:unprocessable_entity)
```
I thought this would also end up in `errors.ex`

mrapacz commented on aion/web/controllers/session_controller.ex:
```diff
@@ -0,0 +1,29 @@
+        |> render("login.json", jwt: jwt)
+      {:error, _reason, conn} ->
+        conn
+        |> put_status(500)
```
Bump

mrapacz commented on aion/web/views/session_view.ex:
```diff
@@ -0,0 +1,11 @@
+    %{"data": "failed to authorise"}
+  end
+
+  def render("login.json", %{jwt: jwt}) do
```
The `jwt` cannot be changed to `token` cause it would require a change in Guardian's implementation, right?

mrapacz commented on aion/web/views/errors.ex:
```diff
@@ -0,0 +1,12 @@
+
+  def unauthenticated(conn) do
+    conn
+      |> put_status(401)
```
The tab here is [redundant](https://github.com/christopheradams/elixir_style_guide#pipe-operator)

mrapacz commented on aion/web/views/errors.ex:
```diff
@@ -0,0 +1,12 @@
+defmodule Aion.ErrorCodesViews do
```
I think it would be easier if the module names resembled the file names, just like they did so far (e.g. RoomChannel -> room_channel.ex, UserRecord -> user_record.ex)

It would be even better if their whole name would represent the filepath, but it's not doable so easily with Phoenix 1.2 architecture. But the file names can mirror the module names at least, it's easier to search through the repo then. 

Also, this is part of a controller, not of a view. I think putting it in the same directory as all the controllers would be alright (the code wouldn't move _that_ far in the whole project directory). You could name it then like `Aion.ControllerErrors` or something similar

mrapacz reviewed (changes requested):
Move the rest of the common errors to the error file. As for the rest of the issues, if something is not to be fixed during this PR just make sure there is an existing issue for that, so that we don't forget to fix it in the future (like the request stuff in Elm) :smiley:

mrapacz committed:
Extract controller errors

mrapacz commented:
I've moved the errors stuff to a separate module in controllers directory. I think it looks better now. Also, the imports used in all of the controllers can be stored in the `web.ex` to avoid repeating stuff.

One last thing that needs to be done is moving Guardian logic to `conn_case.ex`

We've moved to a lib that is widely used so you shouldn't have any problems finding resources in this area like [this](https://elixirforum.com/t/guardian-in-controller-tests/1743) or [this](https://stackoverflow.com/questions/37557737/stub-guardian-plug-ensureauthenticated-for-phoenix-controller-testing)

Commit:
Move setup to conn_case.ex

Commit:
Make two separate ConnCase s

Commit:
Add login tests

Commit:
Refactor login tests

Commit:
Add more test to authorization

pmrukot commented:
@mrapacz Issue with `conn_case` is resolved by adding another test helper since one of our tests need to have different headers to work properly.

I've also added tests for `Aion.Auth` module.

I think we are ready to merge if all issues are resolved.

mrapacz commented on aion/test/auth/auth_test.exs:
```diff
@@ -0,0 +1,45 @@
+
+    login_result = Aion.Auth.login_by_username_and_pass(conn, @user.email, @user.password, repo: Aion.Repo)
+
+    assert tuple_size(login_result) == 2
```
Just for the future, I think in general we could match the exact results (if it's possible), cause just checking for list's/tuple's length may not be the most accurate thing. 

This is what we've been doing so far, so no need to fix it, just suggesting that we could take another approach in the future.

pmrukot replied:
I wanted to do it but it was not possible since authenticated conn contains things that depend on time (I think some kind of timestamp).

Second thing is that the result is a tuple that is `{:ok, login(conn, user)}` and `login` method was tested above. Also its the only result that provides tuple that has length 2 so I think its fine (although not that scalable).

mrapacz commented on aion/test/test_helper.exs:
```diff
@@ -1,19 +1,3 @@
 ExUnit.start
```
I guess this file is not needed anymore?

pmrukot replied:
It is
```ex
19:52:22.599 [error] Task #PID<0.314.0> started from #PID<0.73.0> terminating
** (stop) exited in: GenServer.call(ExUnit.Server, {:take_async_cases, 16}, 60000)
    ** (EXIT) no process: the process is not alive or there's no process currently associated with the given name, possibly because its application isn't started

```

mrapacz reviewed (changes requested):
So if I understand right the AuthConnCase sets up a logged in conn for the tests and ConnCase doesn't? Just small nits, looks good to me apart from that, good job :+1:

mrapacz commented:
Also, make sure you update the description of PR and its title/commit message. As for the latter one - pro tip:
[github browser extensions](https://github.com/collections/github-browser-extensions)
especially [refined github](https://github.com/sindresorhus/refined-github) which will set the commit message automatically as the PR title + issue number

pmrukot commented:
`AuthConnCase` is used used in all controllers tests apart from `PageControllerTest` - it uses old `ConnCase` since it requires different headers to pass and renders html, not json as other controllers.
